Consistent naming of the standard gates

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Ensure consistent naming of the standard gates according to #3633 


### Details and comments

This PR ensures that the names of the standard gates follow the same consistent principles. See #3633 for more detail and the discussion on gate naming. The old/current gates will be deprecated.

Up-to-date list of the names:
```
| file                            | class                    | name           | qc._           |
|---------------------------------+--------------------------+----------------+----------------|
  extensions/standard                 
| barrier.py                      | Barrier                  | barrier        | barrier        |
| ccx.py                          | CCXGate <- ToffoliGate   | ccx            | ccx, toffoli   |
| ch.py                           | CHGate                   | ch             | ch             |
| crx.py                          | CRXGate <- CrxGate       | crx            | crx            |
| cry.py                          | CRYGate <- CryGate       | cry            | cry            |
| crz.py                          | CRZGate <- CrzGate       | crz            | crz            |
| cswap.py                        | CSwapGate <- FredkinGate | cswap          | cswap, fredkin |
| cu1.py                          | CU1Gate <- Cu1Gate       | cu1            | cu1            |
| cu3.py                          | CU3Gate <- Cu3Gate       | cu3            | cu3            |
| cx.py                           | CXGate <- CnotGate       | cx             | cx, cnot       |
| cy.py                           | CYGate <- CyGate         | cy             | cy             |
| cz.py                           | CZGate <- CzGate         | cz             | cz             |
| h.py                            | HGate                    | h              | h              |
| iden.py                         | IGate <- IdGate          | id             | i/id <- iden   |
| ms.py                           | MSGate                   | ms             | ms             |
| r.py                            | RGate                    | r              | r              |
| rx.py                           | RXGate                   | rx             | rx             |
| rxx.py                          | RXXGate                  | rxx            | rxx            |
| ry.py                           | RYGate                   | ry             | ry             |
| rz.py                           | RZGate                   | rz             | rz             |
| rzz.py                          | RZZGate                  | rzz            | rzz            |
| s.py                            | SGate                    | s              | s              |
| s.py                            | SdgGate                  | sdg            | sdg            |
| swap.py                         | SwapGate                 | swap           | swap           |
| t.py                            | TGate                    | t              | t              |
| t.py                            | TdgGate                  | tdg            | tdg            |
| u1.py                           | U1Gate                   | u1             | u1             |
| u2.py                           | U2Gate                   | u2             | u2             |
| u3.py                           | U3Gate                   | u3             | u3             |
| x.py                            | XGate                    | x              | x              |
| y.py                            | YGate                    | y              | y              |
| z.py                            | ZGate                    | z              | z              |
| multi_control_rotation_gates.py |                          |                | mcrx           |
|                                 |                          |                | mcry           |
|                                 |                          |                | mcrz           |
| multi_control_toffoli_gate.py   |                          |                | mct            |
| multi_control_u1_gate.py        |                          |                | mcu1           |
| relative_phase_toffoli.py       |                          |                | rccx           |
|                                 |                          |                | rcccx          |
   extensions/quantum_initializer
| diagonal.py <- diag.py          | DiagonalGate <- DiagGate | diagonal <- diag | diagonal <- diag_gate
| initializer.py                  | Initialize               | initialize     | initialize     |
| mcg_up_to_diagonal.py           | MCGupDiag                | MCGupDiag      |                |
| squ.py                          | SingleQubitUnitary       | unitary        | squ            |
| uc.py <- ucg.py                 | UCGate <- UCG            | multiplexer    | uc <- ucg      |
| uc_pauli_rot.py <- ucrot.py     | UCPauliRotGate <- UCRot  |                |                |
| ucrx.py <- ucx.py               | UCRXGate <- UCX          | ucrx <- ucrotX | ucrx <- ucx    |
| ucry.py <- ucy.py               | UCRYGate <- UCY          | ucry <- ucrotY | ucry <- ucy    |
| ucrz.py <- ucz.py               | UCRZGate <- UCZ          | ucrz <- ucrotZ | ucrz <- ucz    |
| isometry.py                     | Isometry                 | iso            | iso            |
```

Gate checklist:
- [x] Pauli gates and H gate
- [x] identity gate
- [x] T and S gate, including inverses
- [x] multi-controlled gates
- [x] controlled U1/2/3 gates
- [x] controlled swap gate
- [x] diagonal gate
- [x] initializer **Edit:** not a gate since it contains a reset, which is not unitary
- [x] multiplexer 
- [x] uniformly controlled Pauli gates
- [x] isometry **Edit:** not a gate since it doesn't need to be represented via a square matrix
- [x] add note on `QuantumCircuit` method aliases in renos
- [x] update renos on gate names

Note: `MCGupDiag` is left as is, this refactor will be tackled via #3761.

Cryoris commented:
For the deprecation warnings we could have two mechanisms (and probably also others).
The old gate names could be functions returning the new class:
```
def ToffoliGate():
    import warnings
    warnings.warn('ToffoliGate is deprecated, use CCXGate instead!', DeprecationWarning, 2)
    return CCXGate()
```
or we could derive the old class from the new one, to still give access to something of the old-class type:
```
class ToffoliGate(CCXGate):
    def __init__(self):
        import warnings
        warnings.warn('ToffoliGate is deprecated, use CCXGate instead!', DeprecationWarning, 2)
        super().__init__()
```
Latter seems safer to me, since we still have the old class available as type. 

Is there another standard procedure to deprecate class names?

kdk commented:
The overall approach looks good. For the deprecation, we should think through what cases we want to cover, because I agree handling it in full generality wouldn't be straightforward. While the second example supports `isinstance(ToffoliGate(), CCXGate)`, it doesn't support `isinstance(CCXGate(), ToffoliGate)`.

It might be beneficial to split this in two PRs. One to rename and deprecate the to-be-renamed classes, and another to update all the instances where they are used.

Cryoris commented:
> While the second example supports `isinstance(ToffoliGate(), CCXGate)`, it doesn't support `isinstance(CCXGate(), ToffoliGate)`.

That's correct, but this instance-checking is not really something we expect anybody to do, or is it?

ewinston commented:
To pass @kdk's check you may be able to define a metaclass like

```
class ToffoliGateMeta(type):
    def __instancecheck__(self, inst):
        if type(inst) in {CCXGate, ToffoliGate}:
            return True
        return False

class CCXGate(ControlledGate, metaclass=ToffoliGateMeta):
```
...
```
class ToffoliGate(CCXGate, metaclass=ToffoliGateMeta):
    def __init__(self):
        import warnings
        warnings.warn('ToffoliGate is deprecated, use CCXGate instead!', DeprecationWarning, 2)
        super().__init__()
```

Cryoris commented:
Nice workaround (though it seems like a lot of work to deprecate a class). But to ensure everything's running smoothly we probably should implement that.

Cryoris commented:
Updating @kdk's list from #3633, the gate names are now
```
| file                            | class                    | name        | qc._           |
|---------------------------------+--------------------------+-------------+----------------|
  extensions/standard                 
| barrier.py                      | Barrier                  | barrier     | barrier        |
| ccx.py                          | CCXGate <- ToffoliGate   | ccx         | ccx, toffoli   |
| ch.py                           | CHGate                   | ch          | ch             |
| crx.py                          | CRXGate <- CrxGate       | crx         | crz            |
| cry.py                          | CRYGate <- CryGate       | cry         | crz            |
| crz.py                          | CRZGate <- CrzGate       | crz         | crz            |
| cswap.py                        | CSwapGate <- FredkinGate | cswap       | cswap, fredkin |
| cu1.py                          | CU1Gate <- Cu1Gate       | cu1         | cu1            |
| cu3.py                          | CU3Gate <- Cu3Gate       | cu3         | cu3            |
| cx.py                           | CXGate <- CnotGate       | cx          | cx, cnot       |
| cy.py                           | CYGate <- CyGate         | cy          | cy             |
| cz.py                           | CZGate <- CzGate         | cz          | cz             |
| h.py                            | HGate                    | h           | h              |
| iden.py                         | IGate <- IdGate          | i <- id     | i <- iden      |
| ms.py                           | MSGate                   | ms          | ms             |
| r.py                            | RGate                    | r           | r              |
| rx.py                           | RXGate                   | rx          | rx             |
| rxx.py                          | RXXGate                  | rxx         | rxx            |
| ry.py                           | RYGate                   | ry          | ry             |
| rz.py                           | RZGate                   | rz          | rz             |
| rzz.py                          | RZZGate                  | rzz         | rzz            |
| s.py                            | SGate                    | s           | s              |
| s.py                            | SdgGate                  | sdg         | sdg            |
| swap.py                         | SwapGate                 | swap        | swap           |
| t.py                            | TGate                    | t           | t              |
| t.py                            | TdgGate                  | tdg         | tdg            |
| u1.py                           | U1Gate                   | u1          | u1             |
| u2.py                           | U2Gate                   | u2          | u2             |
| u3.py                           | U3Gate                   | u3          | u3             |
| x.py                            | XGate                    | x           | x              |
| y.py                            | YGate                    | y           | y              |
| z.py                            | ZGate                    | z           | z              |
| multi_control_rotation_gates.py |                          |             | mcrx           |
|                                 |                          |             | mcry           |
|                                 |                          |             | mcrz           |
| multi_control_toffoli_gate.py   |                          |             | mct            |
| multi_control_u1_gate.py        |                          |             | mcu1           |
| relative_phase_toffoli.py       |                          |             | rccx           |
|                                 |                          |             | rcccx          |
   extensions/quantum_initializer
| diag.py                         | DiagGate                 | diag        | diag_gate      |
| initializer.py                  | Initialize               | initialize  | initialize     |
| mcg_up_to_diagonal.py           | MCGupDiag                | MCGupDiag   |                |
| squ.py                          | SingleQubitUnitary       | unitary     | squ            |
| uc.py <- ucg.py                 | UCGate <- UCG            | multiplexer | uc <- ucg      |
| ucrot.py                        | UCRot                    |             |                |
| ucrx.py <- ucx.py               | UCRXGate <- UCX          | ucrotX      | ucrx <- ucx    |
| ucry.py <- ucy.py               | UCRYGate <- UCY          | ucrotY      | ucry <- ucy    |
| ucrz.py <- ucz.py               | UCRZGate <- UCZ          | ucrotZ      | ucrz <- ucz    |
| isometry.py                     | Isometry                 | iso         | iso            |
```

ewinston commented:
To add to the gate names above there is also 
UnitaryGate which has name "unitary" and `QuantumCircuit.unitary`
```
| file       | class       | name    | qc._    |
|------------+-------------+---------+---------|
| unitary.py | UnitaryGate | unitary | unitary |
```
So there is a name collision which could be an issue.

Cryoris commented:
From my understanding a `Gate` can only be a unitary operation. What's the reasonable for the `UnitaryGate`?

Cryoris commented:
> Overall looks good. Some comments I have;
> 
> * Since this PR has many user facing changes could you add deprecations to the release notes. The procedure for this is described in "CONTRIBUTING.md".

That's added.

> * In some gate classes you prepended "The" to the docstring. I think I prefer without but it's probably more in line with PEP8 to have a complete sentence. It just could be more consistent.

As you pointed out, the docstrings are updated in #3472 anyways, so we can outsource the fixing of docstrings in a separate issue/PR. That's now addressed by #3705.

1ucian0 commented on qiskit/qasm/libs/qelib1.inc:
```diff
@@ -29,11 +29,11 @@ gate h a { u2(0,pi) a; }
 gate s a { u1(pi/2) a; }
 // Clifford gate: conjugate of sqrt(Z)
-gate sdg a { u1(-pi/2) a; }
+gate sinv a { u1(-pi/2) a; }
```
I dont think the names in the qasm dump should change, since the standard library was no way to deprecated previous names. Maybe create  `qiskit/qasm/libs/qelib2.inc` for that? @ajavadia ?

Cryoris replied:
I changed these lines to
```
gate sdg a { u1(-pi/2) a; }
gate sinv a { sdg a; }
```
That should work, wouldn't it?

ewinston replied:
This does not seem to work with `QuantumCircuit.from_qasm_file`. I get 

> Custom non-opaque gates are not supported by ast_to_dag module

Cryoris replied:
Oh, yes this happened because the `sdg`/`tdg` were changed to `*inv` in the interpreter. I added both names `dg` and `inv` as standard gate names. Should be fixed by 76a2346.

ewinston replied:
Could you also add a test case which would catch issues like this? Also it may be better if ast_to_dag populated `standard_extension` from qelib1.inc, although that could be saved for a future pr.

Cryoris replied:
@ewinston 918c087 implements a test to build a circuit from a qasm file including all gates from qelib1 and compares it against the same circuit set up via circuit methods. That could ideally also be automated by taking the gates from `qelib1.inc` itself. 
I agree ast_to_dag should take the gates from qelib directly, that'd be straightforward for the gate names but probably more difficult to automatically find the corresponding `Gate` for the `standard_extension` dict. Also these gates are not really changing, are they?

Cryoris commented:
I think this is pretty much done as far as the renaming is concerned. I have only two questions left:
1. Should the identity gate really become `IGate/qc.i/'i'` and not `IdGate/qc.id/'id'`? It seems odd to me to name the identity gate `i` and not `id`. 
2. Should we move the gates to `qiskit.circuit.gate` and expose a `qiskit.gate` in this PR or do another one for this?

ewinston commented on test/python/circuit/test_circuit_load_from_qasm.py:
```diff
@@ -48,24 +48,117 @@ def test_qasm_file(self):
 
+    def test_loading_all_qelib1_gates(self):
+        """Test setting up a circuit with all gates defined in qiskit/qasm/libs/qelib1.inc."""
+        qasm_str = '\n'.join(["OPENQASM 2.0;",
```
Could you have just loaded this from all_gates.qasm which you added?

Cryoris replied:
I would prefer to keep it as string in the same function, so we don't have to modify code in several places. The `all_gates.qasm` should've been removed, but slipped through my fingers. 

Could `all_gates.qasm` be useful in other places? Then it would of course change things and keeping it seems good.

Cryoris replied:
I think it's better to have it as string in the same function, so all is kept in one place. The `all_gates.qasm` should've been removed but managed to slip through 😉 

If `all_gates.qasm` could be handy for other tests however, that changes things. Then keeping it seems better.

ewinston replied:
The issue I see is that this seems to add to the number of places where something has to change if qelib1 is updated. To reduce that I think it would be good to either remove it here or remove all_gates.qasm since nothing else seems to use that currently.

Cryoris replied:
Done in e5ef275.

ewinston commented:
>     1. Should the identity gate really become `IGate/qc.i/'i'` and not `IdGate/qc.id/'id'`? It seems odd to me to name the identity gate `i` and not `id`.

That seems to be the consensus.
 
>     2. Should we move the gates to `qiskit.circuit.gate` and expose a `qiskit.gate` in this PR or do another one for this?

Let's save it for a subsequent PR.

ajavadia commented:
I might have missed this but shouldn't `SinvGate` and `TinvGate` be `SInvGate` and `TInvGate`?

kdk commented on qiskit/extensions/standard/t.py:
```diff
@@ -78,20 +89,35 @@ def inverse(self):
 def tdg(self, q):
-    """Apply Tdg to q."""
+    """Apply Tdg (deprecated!) to q."""
+    warnings.warn('tdg() is deprecated, use tinv() instead!', DeprecationWarning, 2)
```
Are `tdg` and `sdg` going to be kept as aliases?

Cryoris replied:
We can also keep them, I don't have such a strong opinion on this. 👍

kdk commented on releasenotes/notes/consistent-gate-naming-efa669c7a8d3a654.yaml:
```diff
@@ -0,0 +1,30 @@
+    UCGate <- UCG            | multiplexer | uc <- ucg
+    UCRXGate <- UCX          | ucrotX      | ucrx <- ucx
+    UCRYGate <- UCY          | ucrotY      | ucry <- ucy
+    UCRZGate <- UCZ          | ucrotZ      | ucrz <- ucz
```
Should S and T gates be included here?

Cryoris replied:
They should and are now!

kdk commented on qiskit/extensions/standard/t.py:
```diff
@@ -78,20 +89,35 @@ def inverse(self):
+    """The deprecated TInv gate."""
+
+    def __init__(self):
+        warnings.warn('TdgGate is deprecated, use TInvGate instead!', DeprecationWarning, 2)
```
Can the deprecation warnings be updated to include the information from https://qiskit.org/documentation/contributing_to_qiskit.html#deprecation-warnings ?

Cryoris replied:
Yes, what version should I indicate? As in  ```... will be deprecated as of 0.x.x```.

kdk replied:
Terra's next minor release is 0.12.0, so we can aim for that.

kdk commented on qiskit/schemas/deprecated/qobj/qobj_generic_schema.json:
```diff
@@ -26,7 +26,7 @@
             "required": ["qubits"],
             "properties": {
-                "name": {"enum": ["reset", "id", "h", "s", "sdg", "t", "tdg", "x", "y", "z"]},
+                "name": {"enum": ["reset", "id", "h", "s", "sinv", "t", "tinv", "x", "y", "z"]},
```
Aer will need to be updated to expect the new gate names (and report them in `basis_gates`). Can you open an issue there?

Cryoris replied:
Qiskit/qiskit-aer#533

Cryoris commented:
> I might have missed this but shouldn't `SinvGate` and `TinvGate` be `SInvGate` and `TInvGate`?

@ajavadia No you haven't, I corrected that in 0a33761. 

> Let's save it for a subsequent PR.

@ewinston Alright, sounds good, 

> That seems to be the consensus.

@ewinston Okay, renaming `id` to `i` is a bit uncomfortable, I'll commit the changes and start a review on places I'm not sure about.

Cryoris commented on qiskit/assembler/disassemble.py:
```diff
@@ -50,7 +50,7 @@ def _experiments_to_circuits(qobj):
                 name = i.name
-                if i.name == 'id':
+                if i.name == 'i':
                     name = 'iden'
```
Should this case be deleted? The name should stay `'i'`, or why was it previously renamed to `'iden'`?

ewinston replied:
Good catch. Yes, looks like this should be deleted since now i.name is the same as the circuit attribute.

Cryoris commented on qiskit/converters/ast_to_dag.py:
```diff
@@ -161,7 +164,7 @@ def _process_bit_id(self, node):
         if node.type == "indexed_id":
```
Is `indexed_id` an indexed identity gate or does id stand for identifier? Should this be changed to `indexed_i`?

ewinston replied:
This is an indexed node used in the dag.

Cryoris commented on qiskit/schemas/backend_properties_schema.json:
```diff
@@ -1,6 +1,6 @@
 {
     "$schema": "http://json-schema.org/draft-04/schema#",
-    "id": "http://www.qiskit.org/schemas/backend_props_schema.json",
+    "i": "http://www.qiskit.org/schemas/backend_props_schema.json",
```
This was probably not meant to change.

Cryoris commented on qiskit/test/mock/backends/almaden/conf_almaden.json:
```diff
@@ -1 +1 @@
-{"qubit_lo_range": [[4.357219891603379, 5.357219891603379], [4.471718605821988, 5.471718605821988], [4.570921129659001, 5.570921129659001], [4.426860381490795, 5.426860381490795], [4.219762317467256, 5.219762317467256], [4.274980354919749, 5.274980354919749], [4.404451150130192, 5.404451150130192], [4.289202317599421, 5.289202317599421], [4.357611487802057, 5.357611487802057], [4.159223127935142, ...
\ No newline at end of file
+{"qubit_lo_range": [[4.357219891603379, 5.357219891603379], [4.471718605821988, 5.471718605821988], [4.570921129659001, 5.570921129659001], [4.426860381490795, 5.426860381490795], [4.219762317467256, 5.219762317467256], [4.274980354919749, 5.274980354919749], [4.404451150130192, 5.404451150130192], [4.289202317599421, 5.289202317599421], [4.357611487802057, 5.357611487802057], [4.159223127935142, ...
```
In all these config JSON files, the name of the identity gate has changed to `'i'`, but the qasm definition within is still
```
gate id a { U(0, 0, 0) a; }
```
do these need to be the same? I.e. change this to 
```
gate i a { U(0, 0, 0) a; }
```

ajavadia replied:
no I think here we are just renaming Terra's user-facing gate names (`IGate` and `circuit.i()`). We are not changing QASM nor backend configurations, which are based on QASM.

Cryoris replied:
Okay, I reverted the backend configs via b625908. One thing that still changed however are the qobj schemata (e.g. `qiskit/schemas/qobj_schema.json`). These are fine to change, right?

Cryoris replied:
Update: keeping the names `sdg`, `tdg` and `id` the schemata do not change.

ewinston commented on qiskit/schemas/backend_configuration_schema.json:
```diff
@@ -1,6 +1,6 @@
 {
     "$schema": "http://json-schema.org/draft-04/schema#",
-    "id": "http://www.qiskit.org/schemas/backend_config_schema.json",
```
Looks like replacement mistake to schema files in this directory.

Cryoris replied:
Yes, already taken care of via 4f8ac24 🙂

Cryoris commented:
Following a discussion on Slack, originating from the troubles of renaming `id/sdg/tdg` to `i/sinv/tinv` in Aer, we'll revert these renamings. For consistency with the other Pauli gates, however, the identity will be the `IGate` with the name `id`. 
Thus:
```
| file            | class                    | name          | qc._           |
|-----------------------------------------------------------------------------|
| i.py            | IGate <- IdGate          | id            | i/id <- iden   |
| s.py            | SdgGate                  | sdg           | sdg            |
| t.py            | TdgGate                  | tdg           | tdg            |
```

ewinston commented on qiskit/converters/ast_to_dag.py:
```diff
@@ -283,7 +283,7 @@ def _process_node(self, node):
 
         elif node.type == "id":
-            raise QiskitError("internal error: _process_node on id")
+            raise QiskitError("internal error: _process_node on i")
```
I went ahead and made this minor correction.

Cryoris replied:
Thanks!

ewinston commented:
Is the plan to keep the QuantumCircuit method aliases (e.g. qc.iden)? I don't see a problem with doing that. Maybe could mention the approach in the release notes.

Cryoris commented:
> Is the plan to keep the QuantumCircuit method aliases (e.g. qc.iden)? I don't see a problem with doing that. Maybe could mention the approach in the release notes.

The aliases are summarized in the table of [the description](https://github.com/Qiskit/qiskit-terra/pull/3695#issue-359984725). These are `toffoli,ccx`, `fredkin,cswap` and `i,id`. The plan is to deprecate the method `iden`, since it doesn't fit the gate naming convention (`i`) or the mathematical convention (`id`).

ewinston commented:
The summary mentions quantum_initializer gates as remaining items but perhaps these can be put into a separate pr since these are not in the standard gates this pr targets and would keep the pr a bit more manageable.

Cryoris commented on releasenotes/notes/consistent-gate-naming-efa669c7a8d3a654.yaml:
```diff
@@ -0,0 +1,30 @@
+    * the Pauli gates all have one uppercase letter only (I, X, Y, Z)
+    * the parameterized Pauli gates (i.e. rotations) prepend the uppercase letter R (RX, RY, RZ)
+    * a controlled version prepends the uppercase letter C (CX, CRX, CCX)
+    * gates are named according to their action, not their alternative names (CCX, not Toffoli)
```
```suggestion
    * gates are named according to their action, not their alternative names known from e.g. literature (hence CCX, not Toffoli)
```

1ucian0 commented on qiskit/extensions/quantum_initializer/uc.py:
```diff
@@ -305,7 +307,17 @@ def ucg(self, gate_list, q_controls, q_target, up_to_diagonal=False):
+    Deprecated version of uc.
+    """
+    import warnings
+    warnings.warn('qc.ucg is deprecated, use qc.uc instead!', DeprecationWarning, 2)
```
This message does not match the style of other deprecation messages.
```suggestion
    warnings.warn('The QuantumCircuit.ucg() method is deprecated as of 0.14.0, and '
                  'will be removed no earlier than 3 months after that release date. '
                  'You should use the QuantumCircuit.uc() method instead.',
                  DeprecationWarning, stacklevel=2)
```

1ucian0 commented on qiskit/extensions/quantum_initializer/ucrx.py:
```diff
@@ -89,7 +89,17 @@ def ucx(self, angle_list, q_controls, q_target):
+    Deprecated version of ucrx.
+    """
+    import warnings
+    warnings.warn('qc.ucx is deprecated, use qc.ucrx instead!', DeprecationWarning, 2)
```
This message does not match the style of other deprecation messages.
```suggestion
    warnings.warn('The QuantumCircuit. ucx() method is deprecated as of 0.14.0, and '
                  'will be removed no earlier than 3 months after that release date. '
                  'You should use the QuantumCircuit. ucrx() method instead.',
                  DeprecationWarning, stacklevel=2)
```

1ucian0 commented on qiskit/extensions/quantum_initializer/ucry.py:
```diff
@@ -86,7 +86,17 @@ def ucy(self, angle_list, q_controls, q_target):
+    Deprecated version of ucry.
+    """
+    import warnings
+    warnings.warn('qc.ucy is deprecated, use qc.ucry instead!', DeprecationWarning, 2)
```
This message does not match the style of other deprecation messages.
```suggestion
    warnings.warn('The QuantumCircuit. ucy() method is deprecated as of 0.14.0, and '
                  'will be removed no earlier than 3 months after that release date. '
                  'You should use the QuantumCircuit. ucry() method instead.',
                  DeprecationWarning, stacklevel=2)
```

1ucian0 commented on qiskit/extensions/quantum_initializer/ucrz.py:
```diff
@@ -87,7 +87,17 @@ def ucz(self, angle_list, q_controls, q_target):
+    Deprecated version of ucrz.
+    """
+    import warnings
+    warnings.warn('qc.ucz is deprecated, use qc.ucrz instead!', DeprecationWarning, 2)
```
This message does not match the style of other deprecation messages.
```suggestion
    warnings.warn('The QuantumCircuit. ucz() method is deprecated as of 0.14.0, and '
                  'will be removed no earlier than 3 months after that release date. '
                  'You should use the QuantumCircuit. ucrz() method instead.',
                  DeprecationWarning, stacklevel=2)
```

1ucian0 commented on qiskit/extensions/standard/iden.py:
```diff
@@ -15,61 +15,9 @@
-
-
-QuantumCircuit.iden = iden
+warnings.warn('This module is deprecated. The IdGate can now be found in i.py',
```
Hmm.. this `warn` seems strange to me. I guess it's for the user doing `from qiskit.extensions import iden`... but why a user would do that?

Cryoris replied:
That's analogous to the deprecation of the controlled gate modules, e.g.: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/extensions/standard/cx.py

1ucian0 commented on qiskit/extensions/standard/ccx.py:
```diff
@@ -13,12 +13,12 @@
-from qiskit.extensions.standard.x import ToffoliGate, ccx
+from qiskit.extensions.standard.x import ToffoliGate, CCXGate, ccx
 
 warnings.warn('This module is deprecated. The ToffoliGate can now be found in x.py',
```
simila to https://github.com/Qiskit/qiskit-terra/pull/3695/files#r379079419

Cryoris commented:
I noticed that the gates in `quantum_initializer` have been renamed without keeping the old class for deprecation purposes. Afaik these gates are only used for the `isometry` though. Should I add deprecations or is that fine?

ajavadia commented:
@Cryoris I think I've seen diagonal gate being used by people, so yeah it would be good to deprecate those too.

Cryoris commented:
@ajavadia I added the previous classes with a deprecation warning. Also I added a test to
* instantiating all renamed old and new classes
* check the old class is a subclass of the new one (to have the same behaviour)
* check that both the old and new class are instances of one another (Kevin's comment)

ajavadia commented:
Thanks for this effort!
